torch/_decomp/decompositions.py

```diff
@@ -1225,6 +1225,100 @@ def cudnn_batch_norm_backward(
     )
 
 
+@register_decomposition(aten._adaptive_avg_pool2d, disable_meta=True)
+def adaptive_avg_pool2d(input: Tensor, output_size: Tuple[int, int]):
+    # Preconditions
+    device = input.device
+    shape = input.shape
+    ndim = len(shape)
+    utils.check(
+        ndim in (3, 4),
+        lambda: f"adaptive_avg_pool2d(): Expected 3D or 4D tensor, but got {ndim}",
+    )
+    for d in input.shape[-2:]:
+        utils.check(
+            d != 0,
+            lambda: "adaptive_avg_pool2d(): Expected input to have non-zero size for "
+            f"non-batch dimensions, but input has shape {tuple(shape)}.",
+        )
+
+    # Optimisation (we should also do this in the kernel implementation)
+    if shape[-2] % output_size[-2] == 0 and shape[-1] % output_size[-1] == 0:
+        stride = tuple(i // o for i, o in zip(shape[-2:], output_size))
+        kernel = tuple(
+            i - (o - 1) * s for i, o, s in zip(shape[-2:], output_size, stride)
+        )
+        return torch.nn.functional.avg_pool2d(input, kernel, stride)
+
+    def start_index(a, b, c):
+        return (a * c) // b
+
+    def end_index(a, b, c):
+        return (((a + 1) * c) / b).ceil().to(a.dtype)
+
+    # Let's assume the reduction we want to apply is to sum all the elements (averaging from this is easy)
+    # Even more, let's assume that we want to just do the 1d case.
+    # The 2d case is recovered by applying the 1d case along two dimensions
+    # The issue here is that we may want to sum segments of different sizes.
+    # What we do is to get the largest segment, and select all the elements from the initial points
+    # up to the max length. Then we zero out the elements that we picked up and were not necessary if there were any such elements
+    # If all the elements have the same length, we compute the average already, otherwise, we return
+    # the sizes of each window, to compute the sizes of the rectrangles at the end.
+    # This function should recover the efficiency of avg_pool2d if the shape does not need the dynamic window shape
+
+    def adaptive_avg_pool1d(x, dim, out_size):
+        assert dim == -2 or dim == -1
+        in_size = x.size(dim)
+
+        orange = torch.arange(out_size, device=device)
+        i0 = start_index(orange, out_size, in_size)
+        # Let length = end_index - start_index, i.e. the length of the pooling kernels
+        # length.max() can be computed analytically as follows:
+        maxlength = in_size // out_size + 1
+        in_size_mod = in_size % out_size
+        # adaptive = True iff there are kernels with different lengths
+        adaptive = not (in_size_mod == 0 or out_size % in_size_mod == 0)
+        if adaptive:
+            maxlength += 1
+        elif in_size_mod == 0:
+            maxlength -= 1
```

Comment:
Deducing this is non-trivial. I have a sketch of a proof that goes as follows:

Denote `in_size` by `x` and `out_size` by `y`. Let `f(n) = floor(n * x / y)` and let `g(n) = ceil(n * x / y)`. We want to compute the maximum of the function `h(n) = g(n+1) - f(n)` for `n = 0, ..., y - 1`. 
By dividing with reminder `x` into `y`, that is, writing `x = k*y + s`, we see that `max_n h(n) = k + max_n g_s(n+1) - f_s(n)` where `g_s(n) = ceil(n * s / y)`, `f_s(n)=floor(n * s / y)`. In other words, we can substitute `x` by `x % y`.
This already gives the `maxlength` value for `s = 0`, that is, when `y` divides `x`. This is the case `in_size_mod == 0` in the code.
The case `s = 1` is also direct, and you see that it's also non-adaptive (the max is constant). This falls in the `out_size % in_size_mod == 0` case.
The general case `out_size % in_size_mod == 0` follows by noting that the output is periodic with period `in_size_mod` in this case.
Then you need to show that in the rest of the cases you indeed get two different values. 
I don't have time now to finish the argument, but it should not be too difficult to finish it from here

Reply:
fwiw, my number theory has always been quite weak, so there might be a much more direct argument looking at generators for multiplicative groups of integers or smth